lnwallet/channel.go

```diff
@@ -1704,6 +1705,134 @@ func (lc *LightningChannel) SignNextCommitment() ([]byte, error) {
+				var onionBlob [lnwire.OnionPacketSize]byte
+				copy(onionBlob[:], htlc.Payload)
+				updates = append(updates, &lnwire.UpdateAddHTLC{
+					ChanID:      chanID,
+					ID:          htlc.Index,
+					Expiry:      htlc.Timeout,
+					Amount:      htlc.Amount,
+					PaymentHash: htlc.RHash,
+					OnionBlob:   onionBlob,
+				})
+			case Fail:
+				updates = append(updates, &lnwire.UpdateFailHTLC{
+					ChanID: chanID,
+					ID:     htlc.Index,
+					Reason: lnwire.OpaqueReason([]byte{}),
+				})
+			case Settle:
+				updates = append(updates, &lnwire.UpdateFufillHTLC{
+					ChanID:          chanID,
+					ID:              htlc.Index,
+					PaymentPreimage: htlc.RPreimage,
+				})
+			}
+		}
+
+		// Generate last sent commit sig message by signing the transaction and
+		// creating the signature.
+		lc.signDesc.SigHashes = txscript.NewTxSigHashes(commitment.txn)
+		sig, err := lc.signer.SignOutputRaw(commitment.txn, lc.signDesc)
+		if err != nil {
+			return nil, err
+		}
+
+		commitSig, err := btcec.ParseSignature(sig, btcec.S256())
+		if err != nil {
+			return nil, err
+		}
+		updates = append(updates, &lnwire.CommitSig{
+			ChanID:    chanID,
+			CommitSig: commitSig,
+		})
+
+	} else if commitmentChainDiff < 0 {
+		// Remote node claims that it received the commit sig message which we
+		// did not send.
+		return nil, errors.New("remote side claims that it haven't received " +
+			"acked commit sig message")
+	}
+
+	revocationsnumberDiff := msg.NextRemoteRevocationNumber - numberRemoteRevocations
+	if revocationsnumberDiff == 0 {
+		// If remote side expects as receive revocation which we already
+		// consider as last, than it means that they aren't received our
+		// last revocation message.
+		revocationMsg, err := lc.generateRevocation(
+			lc.currentHeight-1,
+			msg.NextRemoteRevocationNumber+1)
+		if err != nil {
+			return nil, err
+		}
+		updates = append(updates, revocationMsg)
+	} else if revocationsnumberDiff < 0 {
+		// Remote node claims that it received the revoke_and_ack message
+		// which we did not send.
+		return nil, errors.New("remote side claims that it haven't received " +
+			"acked revoke and ack message")
+	}
+
+	return updates, nil
+}
+
+// LastCounters returns the historical length of the local commimemnt
+// transaction chain and the historical number of the the revocked commiment
+// transactions, whicha are needed in order to generate the channel
+// reestablish message.
+func (lc *LightningChannel) LastCounters() (uint64, uint64) {
+	// As far we store on last commitment transaction we should rely on the
+	// height of the commitment transaction in order to calculate the length.
+	numberLocalCommitments := lc.localCommitChain.tip().height + 1
+
+	// Number of the revocations might be calculated as the height of the
+	// commitment transactions which will be revoked next minus one. And plus
+	// one because height starts from zero.
+	numberRemoteRevocations := lc.remoteCommitChain.tail().height - 1 + 1
```

Comment:
`-1 + 1` is a noop. 

The proper value for this parameter is simply: one the height of the remote party's commitment chain. Unfortunately, the wording within the spec is very verbose and ambiguous, so it obscures this simple fact.